ER Classifiers switched to SparseNetworkLBP

bhargav opened the pull request:
- Deployed new models and tested locally.
- Fixed #327 and #294

JointTraining was not updating weights when SparseNetworkLBP was used due to a stale reference to `network`. Updated the SparseNetworkLBP class to not store any values and interact with the actual class instead.

Changes in saul-core:
- Make the count variable an `AtomicInteger` instead of using `synchronized` to increment/decrement
- `addInstance` method has documentation and change in formatting. No functionality change.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/SparseNetworkLBP.scala:
```diff
@@ -5,20 +5,20 @@ import edu.illinois.cs.cogcomp.lbjava.learn.{ LinearThresholdUnit, SparseNetwork
+    * @return LTU unit of type [[LinearThresholdUnit]]
+    */
+  override def getLTU(i: Int): LinearThresholdUnit = super.getLTU(i).asInstanceOf[LinearThresholdUnit]
 }
```
Would it make sense to remove this file and use the values directly?  We have access to all the possible parameters ...

bhargav replied:
So, `this.conjunctiveLabels` boolean is protected in the base class and we are trying to set the value in the `JointTrainSparseNetworkLBP` methods. I removed other redundant methods.

danyaljj replied:
Why not use this? https://github.com/IllinoisCogComp/lbjava/blob/master/lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SparseNetworkLearner.java#L157

bhargav replied:
So that boolean does not have a public setter, hence we'll have to override it for now. Will see if it would make sense to move the weight update logic inside the LBJava SparseNetworkLearner class and call it directly.

danyaljj replied:
Correct. If you send a PR for it in lbjava, that'd be great.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala:
```diff
@@ -125,25 +126,32 @@ abstract class ConstrainedClassifier[T <: AnyRef, HEAD <: AnyRef](val onClassifi
+          //  If pathToHead is not defined, try getting the node of onClassifier
+          onClassifier match {
+            case clf: Learnable[T] => Some(clf.node)
+            case _ => println("ERROR: pathToHead is not provided and the onClassifier is not a Learnable!"); None
```
logger.error?

danyaljj replied:
and remove "ERROR: "

bhargav replied:
Fixed.

danyaljj commented:
About the changes in results of L+I: I think the [existing evaluation](https://github.com/bhargav/saul/blob/1051b4a4bb196e082d33a89365e13e0f5d20f4e0/saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/EntityRelation/EntityRelationApp.scala#L96-L97) is not quite correct.

Two scenarios: 
 - `constrainedClassifier.test(instanceCollection)`: it will evaluate over all the instances given in the input.  
 - `constrainedClassifier.test()`:  will collect all the instances [using `pathToHead`](https://github.com/bhargav/saul/blob/1051b4a4bb196e082d33a89365e13e0f5d20f4e0/saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala#L136); this doesn't always correspond to the instances we want/should test on. The whole point I was arguing in #294 is that we should never use `pathToHead` when it comes to evaluation (and it should be used only on the inference time). 

I don't remember how we reported the previous results in the readme file, but I think it was using the latter. I think we should update the results after [dropping `pathToHead` from testing](https://github.com/bhargav/saul/blob/1051b4a4bb196e082d33a89365e13e0f5d20f4e0/saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala#L136).

bhargav commented:
@danyaljj In the PR, `deriveTestInstances` method uses `pathToHead` to get the incident `Node` object. Once we have the node, I am using the `getTestingInstance` on the node to retrieve test candidates for the default `test()` method. Can you have a look at the latest iteration? I had already updated results after making those changes.

danyaljj commented:
Oops. All looks good.

bhargav commented:
Updated LBJava Version and removed the `SparseNetworkLBP` file. Updated er-models package. But I see that some tests in POS and SRL are failing. Will investigate later.

danyaljj commented:
Shoot this is a little sad ... the pre-trained models for SRL were trained on `SparseNetworkLBP ` and need to be retrained; @kordjamshidi how hard is re-training `predicateSenseClassifier ` and `argumentTypeLearner ` and `argumentXuIdentifierGivenApredicate `?

bhargav commented:
Yeah. This might have to wait till next week then. I updated the ER Models but other examples need to be re-trained. 

@danyaljj Not sure why POS test is failing though.

kordjamshidi commented:
Yes, this is what I mentioned before, we need to remove this class after the deadline.

danyaljj commented:
@bhargav I think somehow the new lbjava version brings new version of Edison (which has had some changes) and causes some of the feature-extraction related stuff to fail (related to #339).

bhargav commented:
@danyaljj @kordjamshidi Updated this PR. I'll have to train the SRL models again.

kordjamshidi commented:
yes, it would be great to merge this one asap. @bhargav

bhargav commented:
I tried running SRL training a couple of time and end up with Heap Size limit exceeded error while parsing data model. By the time 1k sentences were parsed, the process had used ~35Gb. Training examples have 37k sentences. I think we'll have to look at memory usage and fix issues there first.

kordjamshidi commented:
You can look at it but I could run it by increasing the memory, using the instructions that I put here: 
https://github.com/IllinoisCogComp/saul/wiki/Common-Erros, maybe try this and see if it works for you.

bhargav commented:
@danyaljj @kordjamshidi This is ready. The performance on SRL task was similar. It was less ( -0.5 %) on `aTr` but equal or better `bTr, cTr, dTr, fTr`.

kordjamshidi commented:
ok, great.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala:
```diff
@@ -132,25 +132,28 @@ abstract class ConstrainedClassifier[T <: AnyRef, HEAD <: AnyRef](val onClassifi
   }
 
-  /** Test with given data, use internally
+  /** Derives Test Instances from the data model.
```
minor: `instances` lower case?

danyaljj replied:
And do we always add `.` at the of comments?

bhargav replied:
Updated. We do have `.` at the end of some comments. No strict style guide though.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala:
```diff
@@ -132,25 +132,28 @@ abstract class ConstrainedClassifier[T <: AnyRef, HEAD <: AnyRef](val onClassifi
-
-    val data: List[T] = allHeads.flatMap(getCandidates).toList.distinct
+      })
+      .map(node => node.getTestingInstances)
```
Minor: ` .map(_.getTestingInstances)`?

danyaljj replied:
Same here: keeping it this way might increase readability.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/ConstrainedClassifier.scala:
```diff
@@ -132,25 +132,28 @@ abstract class ConstrainedClassifier[T <: AnyRef, HEAD <: AnyRef](val onClassifi
-    val allHeads: Iterable[HEAD] = {
-      if (pathToHead.isEmpty) {
+  private def deriveTestInstances: Iterable[T] = {
+    pathToHead.map(edge => edge.from)
```
Minor: I usually would say `pathToHead.map(_.from)`, but now seems like keeping it this way would increase the readability.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala:
```diff
@@ -85,42 +87,47 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
+      incoming.foreach(_.populateUsingTo(instance, train))
+    }
+
+    joinNodes.foreach(_.addFromChild(this, instance, train, populateEdge))
```
Could you document the observation that this line takes too much time?

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala:
```diff
@@ -85,42 +87,47 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
+    if (containsNT(nodeInstance)) {
+      //       logger.warn(s"The instance $t is duplicate and it will be ignored! " +
+      //         s"This might be because you add the same instance to both train and test set. ")
+      return
```
[Some scala programmer will curse you](https://tpolecat.github.io/2014/05/09/return.html) one day for this.

bhargav replied:
Removed the `return` statement and make it `logger.trace`. I do not mind having an `else` statement here as the lines are not too long to cause indent issues.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala:
```diff
@@ -275,7 +282,7 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
 
   def loadDerivedInstances(in: ExceptionlessInputStream) = {
-    count = in.readInt()
+    count = new AtomicInteger(in.readInt())
```
Something to remember for future: replacing may of our integers/counters with `AtomicInteger` might be a good idea.

danyaljj commented on saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala:
```diff
@@ -290,10 +297,9 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
 
   def clearPropertyCache(): Unit = {
-    if (propertyCacheList nonEmpty) {
+    if (propertyCacheList.nonEmpty) {
```
👍

danyaljj commented on saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala:
```diff
@@ -15,15 +15,17 @@ class ModelsTest extends FlatSpec with Matchers {
+            case "A0" => (result.f1 >= 0.9) should be(true)
+            case "A1" => (result.f1 >= 0.9) should be(true)
+            case "A2" => (result.f1 >= 0.6) should be(true)
+            case _ => (result.f1 >= 0.0) should be(true)
```
could you [add upperbound](https://github.com/IllinoisCogComp/saul/blob/master/saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/POSTagger/POSTaggerTest.scala#L135) to these F-scores?

bhargav replied:
Updated.

danyaljj commented:
Just a few comments. Feel free to merge after addressing them.

danyaljj commented on saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala:
```diff
@@ -78,11 +80,11 @@ class ModelsTest extends FlatSpec with Matchers {
-          case "A2" => (result.f1 >= 0.6) should be(true)
+          case "A0" => result.f1 should be(0.95 +- 0.05)
+          case "A1" => result.f1 should be(0.95 +- 0.05)
+          case "A2" => result.f1 should be(0.8 +- 0.2)
```
lol this bounds are super-big! could you bring then in the margin of `0.02`?

danyaljj commented on saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala:
```diff
@@ -15,23 +15,25 @@ class ModelsTest extends FlatSpec with Matchers {
-      case result =>
-        result.label match { case "true" => (result.f1 >= 0.98) should be(true) }
+      result =>
+        result.label match { case "true" => result.f1 should be(0.99 +- 0.01) }
```
coolio!

danyaljj commented on saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala:
```diff
@@ -15,23 +15,25 @@ class ModelsTest extends FlatSpec with Matchers {
+            case "A0" => result.f1 should be(0.95 +- 0.05)
+            case "A1" => result.f1 should be(0.95 +- 0.05)
+            case "A2" => result.f1 should be(0.8 +- 0.2)
+            case _ => (result.f1 >= 0.0) should be(true)
```
Same here: decrease the error margin to `0.02`?

danyaljj commented:
Small comment on the test error margin. Otherwise good to merge.

bhargav commented:
Updated test margins. :)

danyaljj commented:
Looks good. Merging!
